### PR TITLE
Piggy back archival uploading progress on heartbeat details

### PR DESCRIPTION
**Pull request by yycptt**
https://github.com/uber/cadence/issues/1536

- [x] Fix and add unit tests
- [x] Clean up code

### CODE REVIEW

**Comment by yycptt** [comment]
Also fixed https://github.com/uber/cadence/issues/1959

File: service/worker/archiver/activities.go [hunk 1]
```diff
@@ -209,28 +250,39 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
...
-			return err
+			return result, err
 		}
 		handledLastBlob = *historyBlob.Header.IsLast
+
+		progress.UploadedBlobs = append(progress.UploadedBlobs, key.String())
+		currIteratorState, err := historyBlobIterator.GetState()
+		if err == nil {
+			progress.IteratorState = currIteratorState
+			progress.BlobPageToken = pageToken
```
Comment (andrewjdawson2016) [inline service/worker/archiver/activities.go]: It feels weird to me that BlobPageToken and UploadedBlobs can become out of sync. The way I am reading this code is that progress.BlobPageToken represents the highest blob page you have successfully uploaded so far. But in the case where historyBlobIterator.GetState() returns error you get a progress which has a blob in UploadedBlobs with a higher page than progress.BlobPageToken

File: service/worker/archiver/activities.go [hunk 2]
```diff
@@ -209,28 +250,39 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
...
+			progress.HandledLastBlob = handledLastBlob
+		} else {
+			logger.Error("failed to get history blob iterator state", tag.Error(err))
+		}
+		activity.RecordHeartbeat(ctx, progress)
```
Comment (andrewjdawson2016) [inline service/worker/archiver/activities.go]: This doesn't return an error?

File: service/worker/archiver/archiver.go [hunk 3]
```diff
@@ -114,7 +115,11 @@ func handleRequest(ctx workflow.Context, logger log.Logger, metricsClient metric
 	}
 	actCtx := workflow.WithActivityOptions(ctx, ao)
 	uploadSW := metricsClient.StartTimer(metrics.ArchiverScope, metrics.ArchiverUploadWithRetriesLatency)
-	err := workflow.ExecuteActivity(actCtx, uploadHistoryActivityFnName, request).Get(actCtx, nil)
+	var result uploadResult
+	err := workflow.ExecuteActivity(actCtx, uploadHistoryActivityFnName, request).Get(actCtx, &result)
+	if err == nil && result.ErrorReason != "" {
+		err = errors.New(result.ErrorReason)
```
Comment (andrewjdawson2016) [inline service/worker/archiver/archiver.go]: ```err = fmt.Errorf("%v: %v", result.ErrorReason, result.ErrorDetails)```

File: service/worker/archiver/archiver.go [hunk 4]
```diff
@@ -123,7 +128,16 @@ func handleRequest(ctx workflow.Context, logger log.Logger, metricsClient metric
 	}
 	uploadSW.Stop()
 
-	if err != nil {
+	blobsToDelete := result.BlobsToDelete
+	if timeoutErr, ok := err.(*workflow.TimeoutError); ok {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/archiver.go]: Can you comment this bit of code, it took me a little while to understand why this exists.

File: service/worker/archiver/history_blob_iterator.go [hunk 5]
```diff
@@ -274,3 +294,14 @@ func (e *jsonSizeEstimator) EstimateSize(v interface{}) (int, error) {
 func NewJSONSizeEstimator() SizeEstimator {
 	return &jsonSizeEstimator{}
 }
+
+// reset resets iterator to a certain state given its encoded representation
+// if it returns an error, the operation will have no effect on the iterator
+func (i *historyBlobIterator) reset(stateToken []byte) error {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/history_blob_iterator.go]: Awesome :)

File: service/worker/archiver/util.go [hunk 6]
```diff
@@ -122,11 +122,29 @@ func historyMutated(historyBlob *HistoryBlob, request *ArchiveRequest) bool {
 func validateArchivalRequest(request *ArchiveRequest) error {
 	if len(request.BucketName) == 0 {
 		// this should not be able to occur, if domain enables archival bucket should always be set
-		return cadence.NewCustomError(errEmptyBucket)
+		return cadence.NewCustomError(errInvalidRequest, errEmptyBucket)
 	}
 	return nil
 }
 
+func getUploadHistoryActivityResponse(progress uploadProgress, err error) (uploadResult, error) {
```
Comment (andrewjdawson2016) [inline service/worker/archiver/util.go]: Lets chat about different approaches to this method. This does not really feel like a utility method to me with how tightly these error types are coupled to the core logic of upload method.

File: service/worker/archiver/activities.go [hunk 7]
```diff
@@ -76,69 +94,92 @@ const (
...
+		} else if err != nil || result.ErrorReason != "" {
+			scope.IncCounter(metrics.ArchiverNonRetryableErrorCount)
 		}
 	}()
-
 	logger := tagLoggerWithRequest(tagLoggerWithActivityInfo(container.Logger, activity.GetInfo(ctx)), request)
+
+	if activity.HasHeartbeatDetails(ctx) {
+		if err := activity.GetHeartbeatDetails(ctx, &progress); err != nil {
+			logger.Info("failed to get previous progress, start from beginning")
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/worker/archiver/activities.go]: Done

File: service/worker/archiver/activities.go [hunk 8]
```diff
@@ -76,69 +94,92 @@ const (
...
 	domainName := domainCacheEntry.GetInfo().Name
 	clusterName := container.ClusterMetadata.GetCurrentClusterName()
+	historyBlobIterator, err := NewHistoryBlobIterator(request, container, domainName, clusterName, progress.IteratorState)
+	if err != nil {
+		logger.Error("failed to decode history blob iterator state, start from beginning", tag.Error(err))
+	}
 	historyBlobReader := container.HistoryBlobReader
 	if historyBlobReader == nil { // only will be set by testing code
-		historyBlobReader = NewHistoryBlobReader(NewHistoryBlobIterator(request, container, domainName, clusterName))
+		historyBlobReader = NewHistoryBlobReader(historyBlobIterator)
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/worker/archiver/activities.go]: Fixed.

File: service/worker/archiver/archiver.go (hunk 4 above)
Comment (yycptt) [reply service/worker/archiver/archiver.go]: Comments added.

File: service/worker/archiver/activities.go [hunk 9]
```diff
@@ -209,28 +250,39 @@ func uploadHistoryActivity(ctx context.Context, request ArchiveRequest) (err err
...
 		if err := uploadBlob(ctx, blobstoreClient, request.BucketName, key, blob); err != nil {
 			logger.Error(uploadErrorMsg, tag.ArchivalUploadFailReason(errorDetails(err)), tag.ArchivalBlobKey(key.String()), tag.Error(err))
-			return err
+			return result, err
 		}
 		handledLastBlob = *historyBlob.Header.IsLast
+
+		progress.UploadedBlobs = append(progress.UploadedBlobs, key.String())
+		currIteratorState, err := historyBlobIterator.GetState()
+		if err == nil {
```
[the remark this answers is not in the document]
Comment (yycptt) [reply service/worker/archiver/activities.go]: Done

File: service/worker/archiver/archiver_test.go [hunk 10]
```diff
@@ -90,13 +94,25 @@ func (s *archiverSuite) TestHandleRequest_UploadFails_NonRetryableError() {
...
 func (s *archiverSuite) TestHandleRequest_UploadFails_ExpireRetryTimeout() {
 	archiverTestMetrics.On("IncCounter", metrics.ArchiverScope, metrics.ArchiverUploadFailedAllRetriesCount).Once()
-	archiverTestMetrics.On("IncCounter", metrics.ArchiverScope, metrics.ArchiverDeleteBlobSuccessCount).Once()
 	archiverTestMetrics.On("IncCounter", metrics.ArchiverScope, metrics.ArchiverDeleteLocalSuccessCount).Once()
 	archiverTestLogger.On("Error", mock.Anything, mock.Anything).Once()
 
 	env := s.NewTestWorkflowEnvironment()
-	env.OnActivity(uploadHistoryActivityFnName, mock.Anything, mock.Anything).Return(workflow.NewTimeoutError(shared.TimeoutTypeStartToClose))
-	env.OnActivity(deleteBlobActivityFnName, mock.Anything, mock.Anything).Return(nil)
+	env.OnActivity(uploadHistoryActivityFnName, mock.Anything, mock.Anything).Return(nil, workflow.NewTimeoutError(shared.TimeoutTypeStartToClose))
```
Comment (yycptt) [inline service/worker/archiver/archiver_test.go]: Will add timeout details after client side test framework is fixed.

File: service/worker/archiver/activities.go (hunk 1 above)
Comment (yycptt) [reply service/worker/archiver/activities.go]: Updated, but I still feel a little bit weird as I think progress.UploadedBlobs should always reflect what has been uploaded. In the case when GetState() returns an error, the blob is still uploaded successfully, so I think progress should still be updated.

**Pull request merged**